addons/binding/org.openhab.binding.edimax/src/main/java/org/openhab/binding/edimax/internal/commands/AbstractCommand.java

```diff
@@ -0,0 +1,360 @@
+
+    /**
+     * Return the extracted return type.
+     *
+     * @param aResponse
+     * @return
+     */
+    protected T getResultValue(String aResponse) {
+        try {
+            String result = extractValueFromXML(aResponse, getXPathString());
+            return unmarshal(result);
+        } catch (XPathExpressionException ex) {
+            ex.printStackTrace();
+        } catch (ParserConfigurationException ex) {
+            ex.printStackTrace();
+        } catch (SAXException ex) {
+            ex.printStackTrace();
+        } catch (IOException ex) {
+            ex.printStackTrace();
+        }
+
+        return null;
+    }
+
+    /**
+     * Returns true if it is a SET command and false for GET.
+     *
+     * @return
+     */
+    protected boolean isSet() {
+        return (setValue != null);
+    }
+
+    /**
+     * Do the command.
+     *
+     * @param ci
+     * @return
+     * @throws IOException
+     */
+    public T executeCommand(ConnectionInformation ci) throws IOException {
+        String lastPart = "smartplug.cgi";
+        String response = executePost(ci.getUrl(), ci.getPort(), lastPart, getCommandString(), ci.getUsername(),
+                ci.getPassword());
+
+        return getResultValue(response);
+    }
+
+    private String executePost(String targetURL, int targetPort, String targetURlPost, String urlParameters,
+            String username, String password) throws IOException {
+        String complete = targetURL + ":" + targetPort + "/" + targetURlPost;
+
+        HttpURLConnection connection = null;
+        try {
+            // Create connection
+            URL url = new URL(complete);
+            connection = (HttpURLConnection) url.openConnection();
+            connection.setRequestMethod("POST");
+
+            connection.setRequestProperty("Connection", "Keep-Alive");
+            connection.setRequestProperty("Content-Type", "application/x-www-form-urlencoded");
+            connection.setRequestProperty("Content-Length", Integer.toString(urlParameters.getBytes().length));
+
+            String userpass = username + ":" + password;
+            String basicAuth = "Basic " + DatatypeConverter.printBase64Binary(userpass.getBytes());
+            connection.setRequestProperty("Authorization", basicAuth);
+
+            connection.setUseCaches(false);
+            connection.setDoOutput(true);
+
+            // Send request
+            DataOutputStream wr = new DataOutputStream(connection.getOutputStream());
+            wr.write(urlParameters.getBytes());
+            wr.close();
+
+            // Get Response
+            InputStream is = connection.getInputStream();
+            BufferedReader rd = new BufferedReader(new InputStreamReader(is));
+            StringBuilder response = new StringBuilder();
+            String line;
+            while ((line = rd.readLine()) != null) {
+                response.append(line);
+                response.append('\r');
+            }
```

Comment:
I believe there are some streams, readers which are (or might) not properly closed (especially if errors occur), getting things like this correct is hard. 

That is one of the reasons why ESH contains some functionality for this, see the method [HttpUtil.executeUrl](https://www.eclipse.org/smarthome/documentation/javadoc/org/eclipse/smarthome/io/net/http/HttpUtil.html#executeUrl-java.lang.String-java.lang.String-java.util.Properties-java.io.InputStream-java.lang.String-int-) can you take a look at that.

As far as I can spot now it should have all you need, if you need any help with just give a shout.

Reply:
I agree - please use HttpUtil instead of your own implementation.